File: cube_voxel.py
"""
CubeVoxel Class Implementation

Description:
    A class representing a voxel (3D pixel) in the form of a cube. This class provides methods
    for translating, scaling, rotating, and applying custom transformations to the cube voxel.
    Rotations can be performed around the voxel's center or an optional external center.

Author:
    JessyJP (2024)

License:
    GPLv3 - This code is released under the GNU General Public License v3.0.
    See LICENSE.md for more details.
"""

from collections import OrderedDict
import numpy as np
from copy import deepcopy

## TODO: it would be good to have unit testing for this class

class CubeVoxel:
    rotation_matrix_cache = OrderedDict()  # Static cache for rotation matrices, using OrderedDict
    cache_limit = 1000  # Limit for the cache size
    
    def __init__(self, id, nodesR=None, faceConnectNodes=None, colourF=[1.0, 1.0, 1.0], colourE=[0, 0, 0], faceAlpha=1.0):
        """
        Initializes a CubeVoxel instance with specified properties or defaults.

        Parameters:
        - id (int): Identifier for the voxel.
        - nodesR (list, optional): List of node positions defining the voxel. Defaults to a unit cube.
        - faceConnectNodes (list, optional): Connectivity list defining the faces of the voxel. Defaults to a unit cube.
        - colourF (list, optional): Face color of the voxel. Defaults to white.
        - colourE (list, optional): Edge color of the voxel. Defaults to black.
        - faceAlpha (float, optional): Transparency of the voxel faces. Defaults to 1.0 (opaque).
        """
        if nodesR is None:
            nodesR = [[0, 0, 0], [0, 0, 1], [1, 0, 1], [1, 0, 0],
                      [0, 1, 0], [0, 1, 1], [1, 1, 1], [1, 1, 0]]
        if faceConnectNodes is None:
            faceConnectNodes = [[1, 4, 3, 2], [5, 8, 7, 6], [1, 2, 6, 5],
                                [3, 4, 8, 7], [2, 3, 7, 6], [1, 5, 8, 4]]
            
        self.id = id
        self.nodesR = np.array(nodesR, dtype=float)
        self.faceConnectNodes = np.array(faceConnectNodes) - 1  # Adjusting for Python's 0-indexing
        self.faceConnectNodes_Fast = self.faceConnectNodes # For fast render
        self.colourF = np.array(colourF)
        self.colourE = np.array(colourE)
        self.faceAlpha = faceAlpha
        self.center = np.mean(self.nodesR, axis=0)
    
    def copy(self):
        """
        Creates a deep copy of the CubeVoxel instance.

        Returns:
        - CubeVoxel: A new instance of CubeVoxel with identical properties.
        """
        # Use deepcopy to ensure all mutable objects are also copied.
        return deepcopy(self)

    # Modification methods

    def translate(self, translation_vector):
        """
        Translates the voxel by a given vector.

        Parameters:
        - translation_vector (list or np.array): The vector by which to translate the voxel.
        """
        self.nodesR += np.array(translation_vector)
        self.center = np.mean(self.nodesR, axis=0)

    def scale(self, scale_factors):
        """
        Scales the voxel by given scale factors along each axis.

        Parameters:
        - scale_factors (list or np.array): Scale factors for the x, y, and z axes.
        """
        self.nodesR *= np.array(scale_factors)
        self.center = np.mean(self.nodesR, axis=0)

    def rotate(self, angle, axis='z', center_of_rotation=None):
        """
        Rotates the voxel around a specified axis by a given angle. Optionally, a center of rotation can be specified.

        Parameters:
        - angle (float): Rotation angle in degrees.
        - axis (str): Axis of rotation ('x', 'y', or 'z').
        - center_of_rotation (list or np.array, optional): The center point around which to rotate. Defaults to the voxel's center.
        """
        angle_rad = np.radians(angle)
        rotation_center = self.center if center_of_rotation is None else np.array(center_of_rotation)
        rotation_matrix = CubeVoxel.get_rotation_matrix(angle_rad, axis)

        self.nodesR = np.dot(self.nodesR - rotation_center, rotation_matrix) + rotation_center
        self.center = np.mean(self.nodesR, axis=0)

    @staticmethod
    def get_rotation_matrix(angle_rad, axis):
        # Generate a unique key for the cache based on angle and axis
        cache_key = (round(angle_rad, 5), axis)  # Round to handle floating-point precision
        if cache_key in CubeVoxel.rotation_matrix_cache:
            # Move the used key to the end to show it's recently accessed
            CubeVoxel.rotation_matrix_cache.move_to_end(cache_key)
            return CubeVoxel.rotation_matrix_cache[cache_key]
        
        # Compute the rotation matrix if not in cache
        rotation_matrix = CubeVoxel._compute_rotation_matrix(angle_rad, axis)

        # Add the new rotation matrix to the cache
        CubeVoxel.rotation_matrix_cache[cache_key] = rotation_matrix
        # Ensure the cache does not exceed the specified limit
        if len(CubeVoxel.rotation_matrix_cache) > CubeVoxel.cache_limit:
            CubeVoxel.rotation_matrix_cache.popitem(last=False)  # Remove the oldest item
        return rotation_matrix

    @staticmethod
    def _compute_rotation_matrix(angle_rad, axis):
        # Your existing logic to compute the rotation matrix based on the axis
        if axis == 'x':
            return np.array([[1, 0, 0],
                             [0, np.cos(angle_rad), -np.sin(angle_rad)],
                             [0, np.sin(angle_rad), np.cos(angle_rad)]])
        elif axis == 'y':
            return np.array([[np.cos(angle_rad), 0, np.sin(angle_rad)],
                             [0, 1, 0],
                             [-np.sin(angle_rad), 0, np.cos(angle_rad)]])
        elif axis == 'z':
            return np.array([[np.cos(angle_rad), -np.sin(angle_rad), 0],
                             [np.sin(angle_rad), np.cos(angle_rad), 0],
                             [0, 0, 1]])
        else:
            raise ValueError("Axis must be 'x', 'y', or 'z'.")

    def transform_element(self, transformation_function, S):
        """
        Applies a transformation function to the entire voxel.

        Parameters:
        - transformation_function (callable): A function that takes and returns an np.array of node positions.
        """
        self.nodesR = transformation_function(self.nodesR, S)
        self.center = np.mean(self.nodesR, axis=0)

    def transform_node(self, transformation_function, S):
        """
        Applies a transformation function to each node of the voxel individually.

        Parameters:
        - transformation_function (callable): A function that takes and returns a single node position.
        """
        for i, node in enumerate(self.nodesR):
            self.nodesR[i] = transformation_function(node, S)
        self.center = np.mean(self.nodesR, axis=0)

    # Overload the arithmetic operators for easy linear matrix manipulation 

    def __mul__(self, other):
            """
            Supports multiplication of the CubeVoxel with a transformation matrix.

            Parameters:
            - other (np.array): A 3x3 or 4x4 transformation matrix for linear transformations.

            Returns:
            - CubeVoxel: A new instance of CubeVoxel with transformed nodes.
            """
            if not isinstance(other, np.ndarray) or other.shape not in [(3, 3), (4, 4)]:
                raise ValueError("Multiplication is supported only with a 3x3 or 4x4 transformation matrix.")

            # If a 3x3 matrix is provided, apply it directly.
            # If a 4x4 matrix is provided, assume it includes translation and apply it to homogeneous coordinates.
            if other.shape == (3, 3):
                transformed_nodes = np.dot(self.nodesR, other.T)
            else:  # other.shape == (4, 4)
                # Convert nodes to homogeneous coordinates, apply transformation, and convert back.
                nodes_homogeneous = np.hstack([self.nodesR, np.ones((self.nodesR.shape[0], 1))])
                transformed_nodes_homogeneous = np.dot(nodes_homogeneous, other.T)
                transformed_nodes = transformed_nodes_homogeneous[:, :3] / transformed_nodes_homogeneous[:, [3]]

            # Create a new CubeVoxel instance with the transformed nodes.
            return CubeVoxel(self.id, transformed_nodes, self.faceConnectNodes, self.colourF, self.colourE, self.faceAlpha)
    
    def _translate(self, vector, operation):
        """
        A helper method to support translation of the CubeVoxel by a vector.

        Parameters:
        - vector (np.array): A translation vector.
        - operation (callable): The numpy operation to apply (np.add or np.subtract).

        Returns:
        - CubeVoxel: A new instance of CubeVoxel with translated nodes.
        """
        if not isinstance(vector, (np.ndarray, list, tuple)) or len(vector) != 3:
            raise ValueError("Translation is supported only with a 3-element vector.")

        # Apply the translation operation to the nodes.
        translated_nodes = operation(self.nodesR, np.array(vector))

        # Create a new CubeVoxel instance with the translated nodes.
        return CubeVoxel(self.id, translated_nodes, self.faceConnectNodes, self.colourF, self.colourE, self.faceAlpha)

    def __add__(self, other):
        """
        Supports addition of the CubeVoxel with a translation vector.
        """
        return self._translate(other, np.add)

    def __sub__(self, other):
        """
        Supports subtraction of the CubeVoxel with a translation vector.
        """
        return self._translate(other, np.subtract)

File: CubeVoxel.m
classdef CubeVoxel < handle
    % CubeVoxel Class Implementation
    % A class representing a voxel (3D pixel) in the form of a cube.
    % This class provides methods for translating, scaling, rotating,
    % and applying custom transformations to the cube voxel.
    
    properties(Access=public)
        id % Identifier for the voxel
        nodesR % List of node positions defining the voxel
        FaceConnectNodes % Connectivity list defining the faces of the voxel
        colourF = [1.0, 1.0, 1.0] % Face color of the voxel
        colourE = [0, 0, 0] % Edge color of the voxel
        faceAlpha = 1.0 % Transparency of the voxel faces
        edgeAlpha = 1.0 % Transparency of the voxel faces
        center % Center of the voxel
        FaceConnectNodes_Fast;
        offsetNodeInd = 0;
    end
    
    methods
        function obj = CubeVoxel(id, nodesR, FaceConnectNodes, colourF, colourE, faceAlpha)
            % Constructor for CubeVoxel class with default parameters
            
            % Default parameters
            defaultNodesR = [0, 0, 0; 0, 0, 1; 1, 0, 1; 1, 0, 0; ...
                             0, 1, 0; 0, 1, 1; 1, 1, 1; 1, 1, 0];
            defaultFaceConnectNodes = [1, 4, 3, 2; 5, 8, 7, 6; 1, 2, 6, 5; ...
                                       3, 4, 8, 7; 2, 3, 7, 6; 1, 5, 8, 4];
            defaultColourF = [204 204 0]/255;
            defaultColourE = [100 100 100]/255;
            defaultFaceAlpha = 1.0;
            
            % Assign provided parameters or use defaults
            obj.id = id;
            obj.nodesR = defaultNodesR;
            obj.FaceConnectNodes = defaultFaceConnectNodes;
            obj.colourF = defaultColourF;
            obj.colourE = defaultColourE;
            obj.faceAlpha = defaultFaceAlpha;
            
            if nargin >= 2 && ~isempty(nodesR)
                obj.nodesR = nodesR;
            end
            if nargin >= 3 && ~isempty(FaceConnectNodes)
                obj.FaceConnectNodes = FaceConnectNodes;
            end
            if nargin >= 4 && ~isempty(colourF)
                obj.colourF = colourF;
            end
            if nargin >= 5 && ~isempty(colourE)
                obj.colourE = colourE;
            end
            if nargin >= 6 && ~isempty(faceAlpha)
                obj.faceAlpha = faceAlpha;
            end
            
            % Calculate the center based on the nodes
            obj.center = mean(obj.nodesR, 1);
        end
        
        function obj = translate(obj, translation_vector)
            % Translates the voxel by a given vector
            obj.nodesR = bsxfun(@plus, obj.nodesR, translation_vector);
            obj.center = mean(obj.nodesR, 1);
        end
        
        function obj = scale(obj, scale_factors, centerPoint)
            % Scales the voxel by given scale factors along each axis
            % Optionally, scales around a specified center point.
            %
            % Parameters:
            %   scale_factors: A 3-element vector specifying scale factors along each axis (x, y, z).
            %   centerPoint (optional): A 3-element vector specifying the center point to scale around.
            %                           If not provided, the voxel's current center is used.
            
            if nargin < 3
                % If centerPoint is not provided, use the voxel's current center
                centerPoint = obj.center;
            end
            
            % Ensure centerPoint is a row vector for the subtraction and addition operations
            if iscolumn(centerPoint)
                centerPoint = centerPoint';
            end
            
            % Perform scaling around the specified center
            obj.nodesR = bsxfun(@minus, obj.nodesR, centerPoint); % Translate to origin based on centerPoint
            obj.nodesR = bsxfun(@times, obj.nodesR, scale_factors); % Scale
            obj.nodesR = bsxfun(@plus, obj.nodesR, centerPoint); % Translate back to original position based on centerPoint
            
            % Update the center based on the new nodesR
            obj.center = mean(obj.nodesR, 1);
        end
       
        function obj = rotate(obj, angle, axis, pivotPoint)
            % Rotates the voxel around a specified axis by a given angle
            % angle is in degrees
            % pivotPoint is an optional parameter specifying the pivot point for rotation
            
            if nargin < 4
                pivotPoint = obj.center; % Use the voxel's center as the default pivot point
            end
            
            angle_rad = deg2rad(angle); % Convert angle to radians
            rotation_matrix = eye(3); % Initialize rotation matrix
            
            % Define rotation matrices for each axis
            switch axis
                case 'x'
                    rotation_matrix = [1, 0, 0; 0, cos(angle_rad), -sin(angle_rad); 0, sin(angle_rad), cos(angle_rad)];
                case 'y'
                    rotation_matrix = [cos(angle_rad), 0, sin(angle_rad); 0, 1, 0; -sin(angle_rad), 0, cos(angle_rad)];
                case 'z'
                    rotation_matrix = [cos(angle_rad), -sin(angle_rad), 0; sin(angle_rad), cos(angle_rad), 0; 0, 0, 1];
                otherwise
                    error('Axis must be ''x'', ''y'', or ''z''.');
            end
            
            % Translate nodes to the origin based on pivotPoint, apply rotation, and translate back
            obj.nodesR = bsxfun(@minus, obj.nodesR, pivotPoint) * rotation_matrix; % Apply rotation
            obj.nodesR = bsxfun(@plus, obj.nodesR, pivotPoint); % Translate back
            
            % Update the center based on the new nodesR
            obj.center = mean(obj.nodesR, 1);
        end
        
        % Additional custom transformation methods can be added here

        function obj = updateColourPropertiesFrom(obj, otherVoxel)
            % Updates the colour properties from another CubeVoxel instance.
            %
            % Parameters:
            %   otherVoxel: Another instance of CubeVoxel from which to copy
            %   the colour properties.
            
            if nargin < 2 || ~isa(otherVoxel, 'CubeVoxel')
                error('The input must be an instance of CubeVoxel.');
            end
            
            % Update the colour properties
            obj.colourF = otherVoxel.colourF;
            obj.colourE = otherVoxel.colourE;
            obj.faceAlpha = otherVoxel.faceAlpha;
        end
        
        function newObj = deepCopy(obj)
            % Creates a deep copy of the CubeVoxel instance.
            newObj = CubeVoxel(0); % Use the default constructor
            
            % Manually copy all properties
            p = properties(obj);
            for i = 1:length(p)
                newObj.(p{i}) = obj.(p{i});
            end
        end
    end

    methods
        function objString = toOBJString(obj)
            % Generates an OBJ format string for the voxel.
            objString = "";
            for i = 1:size(obj.nodesR, 1)
                objString = objString + sprintf('v %f %f %f\n', obj.nodesR(i, :));
            end
            for i = 1:size(obj.FaceConnectNodes, 1)
                % OBJ indices are 1-based
                indices = obj.FaceConnectNodes(i, :);
                indices = indices + obj.offsetNodeInd;
                % Create a string for the face indices
                faceStr = sprintf('%d ', indices);
                % Trim the last space and add to objString
                objString = objString + sprintf('f %s\n', strtrim(faceStr));
            end
        end

        function stlString = toSTLString(obj)
            % Generates an STL format string for the voxel.
            stlString = "solid voxel\n";
            for i = 1:size(obj.FaceConnectNodes, 1)
                vertices = obj.nodesR(obj.FaceConnectNodes(i, :), :);
                % Assuming calculateNormal can handle more than 3 vertices
                % and returns a proper normal for the polygon
                normal = obj.calculateNormal(vertices);
                % Convert polygon face to triangles and add to STL string
                stlString = stlString + obj.polygonToTrianglesSTL(vertices, normal);
            end
            stlString = stlString + "endsolid voxel\n";
        end

        function trianglesSTL = polygonToTrianglesSTL(~, vertices, normal)
            % This function converts a polygon into a series of triangles
            % following the specified algorithm and returns STL string
            % representation of these triangles.
            trianglesSTL = "";
            if size(vertices, 1) < 3
                error('Not enough vertices to form a polygon.');
            end
            
            for i = 3:size(vertices, 1)
                trianglesSTL = trianglesSTL + sprintf('facet normal %f %f %f\n', normal);
                trianglesSTL = trianglesSTL + "    outer loop\n";
                % First vertex
                trianglesSTL = trianglesSTL + sprintf('        vertex %f %f %f\n', vertices(1, :));
                % (i-1)th vertex
                trianglesSTL = trianglesSTL + sprintf('        vertex %f %f %f\n', vertices(i-1, :));
                % ith vertex
                trianglesSTL = trianglesSTL + sprintf('        vertex %f %f %f\n', vertices(i, :));
                trianglesSTL = trianglesSTL + "    endloop\nendfacet\n";
            end
        end
            
        
       function normal = calculateNormal(~, vertices)
            % Calculates the normal vector for a face given its vertices.
            % Assumes the vertices are provided in a counterclockwise order.
            
            % Calculate vectors along two edges of the triangle
            edge1 = vertices(2, :) - vertices(1, :);
            edge2 = vertices(3, :) - vertices(1, :);
            
            % The normal vector is perpendicular to both edges
            normal = cross(edge1, edge2);
            
            % Normalize the vector to have a unit length
            normal = normal / norm(normal);
        end
    end
end
